src/Rendering/outlineRenderer.ts

```diff
@@ -274,9 +278,44 @@ export class OutlineRenderer implements ISceneComponent {
         // Outline - step 1
         this._savedDepthWrite = this._engine.getDepthWrite();
         if (mesh.renderOutline) {
+            // Cache stencil state
+            var previousStencilBuffer = this._engine.getStencilBuffer();
+            var previousStencilFunction = this._engine.getStencilFunction();
+            var previousStencilMask = this._engine.getStencilMask();
+            var previousStencilOperationPass = this._engine.getStencilOperationPass();
+            var previousStencilOperationFail = this._engine.getStencilOperationFail();
+            var previousStencilOperationDepthFail = this._engine.getStencilOperationDepthFail();
+            var previousStencilReference = this._engine.getStencilFunctionReference();
+
+            var material = subMesh.getMaterial();
+            if (material && material.needAlphaBlending) {
+                // Draw only to stencil buffer for the original mesh
+                // The resulting stencil buffer will be used so the outline is not visible inside the mesh when the mesh is transparent
+                this._engine.setDepthWrite(false);
+                this._engine.setColorWrite(false);
+                this._engine.setStencilBuffer(true);
+                this._engine.setStencilOperationPass(Constants.REPLACE);
+                this._engine.setStencilFunction(Constants.ALWAYS);
+                this._engine.setStencilMask(OutlineRenderer._StencilReference);
+                this._engine.setStencilFunctionReference(OutlineRenderer._StencilReference);
+                this.render(subMesh, batch, /* This sets offset to 0 */ true);
+            }
+
+            // Draw the outline using the above stencil if needed to avoid drawing within the mesh
+            this._engine.setColorWrite(true);
             this._engine.setDepthWrite(false);
+            this._engine.setStencilFunction(Constants.NOTEQUAL);

```

Comment:
what if the mesh is not transparent? do we need all that stencil stuf fin that case?

Reply:
nope, if its not transparent its fine to draw behind

Reply:
so we should do nothing with the stencil then in that case